Release 2

AndrewRadev opened the pull request:
This PR is intended to bring the main repository up to date with the deployed code, and there's been a lot of changes since the last one, #33. I will list high-level changes on the basis of PRs, though there's probably more changes in the "various improvements" categories:

- Scripts to sync data with NCBI (JensenLab, really) and ChEBI: 
    * https://github.com/AndrewRadev/mgrowthdb/pull/55. 
    * *Future updates*: Instead of JensenLab, we need an up-to-date selection of microbial strains. We also should save more metadata for each strain like its parent species and genera. That way, people would be able to search for studies by species or genus.
- Docker setup for development purposes, e.g. for people who want to run and try the app on windows: 
    * https://github.com/AndrewRadev/mgrowthdb/pull/58
- API for muGrowthCtrl to access:
    * https://github.com/AndrewRadev/mgrowthdb/pull/69
    * https://github.com/AndrewRadev/mgrowthdb/pull/73
    * https://github.com/AndrewRadev/mgrowthdb/pull/93
    * https://github.com/AndrewRadev/mgrowthdb/pull/94
- Upload process improvements:
    * https://github.com/AndrewRadev/mgrowthdb/pull/57
    * https://github.com/AndrewRadev/mgrowthdb/pull/78
    * https://github.com/AndrewRadev/mgrowthdb/pull/91
    * https://github.com/AndrewRadev/mgrowthdb/pull/95
- Internal refactoring: Extract "study techniques" as parents to "measurement techniques". This enables us to add labels to techniques and separate live/dead/total cell measurements
    * https://github.com/AndrewRadev/mgrowthdb/pull/79
- New search interface: 
    * https://github.com/AndrewRadev/mgrowthdb/pull/79 
    * https://github.com/AndrewRadev/mgrowthdb/pull/86
- Custom model uploads: 
    * https://github.com/AndrewRadev/mgrowthdb/pull/84 
    * https://github.com/AndrewRadev/mgrowthdb/pull/85
- Sandbox page: 
    * https://github.com/AndrewRadev/mgrowthdb/pull/87 
    * *Future work*: Need to actually make it possible to plot with existing data and fit models. For that purpose, will need to have some temporary storage. Also have an idea for how to integrate it with muGrowthCtrl to let that app push data to the sandbox and visualize it using the API.
- Page visit counting. This should give us an idea of the relative popularity of the project.
    * https://github.com/AndrewRadev/mgrowthdb/pull/88
- Publication authorship. Pretty important to have clear authorship of the studies that are not ours.
    * https://github.com/AndrewRadev/mgrowthdb/pull/90
- Chart improvements: https://github.com/AndrewRadev/mgrowthdb/pull/96
- A lot of smaller improvements that I can't collect easily, but this is the list:
    * https://github.com/AndrewRadev/mgrowthdb/pull/60
    * https://github.com/AndrewRadev/mgrowthdb/pull/63
    * https://github.com/AndrewRadev/mgrowthdb/pull/72
    * https://github.com/AndrewRadev/mgrowthdb/pull/75
    * https://github.com/AndrewRadev/mgrowthdb/pull/80

AndrewRadev committed:
Chart resize fixes

AndrewRadev committed:
Show error message on custom upload form

AndrewRadev committed:
Add description lines to help topics

AndrewRadev committed:
Move help files, rename "Data Analysis"

AndrewRadev committed:
Add KaTeX files to the project to render math

AndrewRadev committed:
Reorganize help pages, add modeling techniques page

AndrewRadev committed:
Only load KaTeX CSS on help page

AndrewRadev committed:
In dev mode, rebuild help files on every load

AndrewRadev committed:
Document new search form

AndrewRadev committed:
Fix help page bug

AndrewRadev committed:
Update screenshots, reorganize export and modeling sections

AndrewRadev committed:
Fix "Add clause" button

AndrewRadev committed:
Add model info in search results

AndrewRadev committed:
New search pagination

AndrewRadev committed:
Swap new search and old ("advanced") search

AndrewRadev committed:
New search <-> advanced search template renaming and fixes

AndrewRadev committed:
Fix template name

AndrewRadev committed:
Advanced search fixes

AndrewRadev committed:
Fix search page tests

AndrewRadev committed:
Minor technique name fix

AndrewRadev committed:
Move help pages back under pages/help, rename to "topics"

AndrewRadev committed:
Format experiment descriptions, document it

AndrewRadev committed:
Fix search documentation

AndrewRadev committed:
Modeling documentation

AndrewRadev committed:
Extract dedicated measurement technique help page

AndrewRadev committed:
Documentation topic rename

AndrewRadev committed:
Add growth phases documentation

AndrewRadev committed:
Unify style of basic and advanced search results

AndrewRadev committed:
Fix tests

AndrewRadev committed:
Even smaller info buttons

AndrewRadev committed:
Adjust UI for experiment descriptions

AndrewRadev committed:
Try different parameters for model-fitting process

AndrewRadev committed:
Merge pull request #84 from AndrewRadev/custom-model-upload

Custom model upload

AndrewRadev committed:
Add output schemas to the API documentation

AndrewRadev committed:
Allow dev documentation updates from this branch

AndrewRadev committed:
Use Easy linear method for a B-R starting point instead of splines

AndrewRadev committed:
Use EL method for an initial logistic fit as well

AndrewRadev committed:
Starting page for the data sandbox

AndrewRadev committed:
More title adjustments

AndrewRadev committed:
Move admin link back to the bottom

It sort of gets in the way, visually

AndrewRadev committed:
Remove usage of `legend_position`

We always put the legend on top now.

AndrewRadev committed:
Load plotly assets in the page head

Previously, this worked at the bottom of the page, because all charts
were being loaded with ajax. Now, we want a page that loads a chart
immediately.

AndrewRadev committed:
Working sandbox chart form without units

AndrewRadev committed:
Upload files, add units

AndrewRadev committed:
Help information

AndrewRadev committed:
Implement "log view" buttons

AndrewRadev committed:
Add label to downloaded CSV files

AndrewRadev committed:
Minor javascript fixes

AndrewRadev committed:
Validate input files on the sandbox

AndrewRadev committed:
Fix log view on sandbox page

AndrewRadev committed:
Basic page view tracking

AndrewRadev committed:
Store additional information for page visits

AndrewRadev committed:
Aggregate page visits

AndrewRadev committed:
Start celery scheduler in prod server file

AndrewRadev committed:
Fix celery scheduler log level

AndrewRadev committed:
Extract sidebar links to a separate template

AndrewRadev committed:
Change sandbox emoji

AndrewRadev committed:
Input for search items per page

AndrewRadev committed:
Merge branch 'documentation-updates' into sandbox-page

* documentation-updates:
  Input for search items per page

AndrewRadev committed:
Merge branch 'sandbox-page' into page-visit-tracking

* sandbox-page:
  Input for search items per page
  Change sandbox emoji
  Extract sidebar links to a separate template

AndrewRadev committed:
Merge pull request #86 from AndrewRadev/documentation-updates

Documentation updates, new search

AndrewRadev committed:
Add some example files to try out sandbox

AndrewRadev committed:
Merge branch 'sandbox-page' into page-visit-tracking

* sandbox-page:
  Add some example files to try out sandbox

AndrewRadev committed:
Set remote address correctly in prod

AndrewRadev committed:
Try a second level of IP-forwarding

IP might be forwarded by KU Leuven as well.

AndrewRadev committed:
Read IP forwarding levels from env file

AndrewRadev committed:
Remove handled todo

AndrewRadev committed:
Merge branch 'sandbox-page' into page-visit-tracking

* sandbox-page:
  Remove handled todo

AndrewRadev committed:
Add country to page views (WIP)

AndrewRadev committed:
Fix javascript bug

AndrewRadev committed:
Order admin listings recent-first, tweak PageVisit rendering

AndrewRadev committed:
Find visitor country based on IP

AndrewRadev committed:
Add filters on page visit admin page

AndrewRadev committed:
Aggregate page views by country

AndrewRadev committed:
Ensure maxminddb lookups don't break the app

AndrewRadev committed:
Parse IPv6 addresses with square brackets

AndrewRadev committed:
Add authorship fields to studies

AndrewRadev committed:
Change authorship columns, add script that fetches info

AndrewRadev committed:
Show authors in search, apply query to them (WIP)

Right now, this only works for last names, because what's actually being
queried is the citation. It might be better to cache an "authorText"
column for querying purposes.

AndrewRadev committed:
Test author name generation and searching

AndrewRadev committed:
Show publication authors on study and experiment pages

AndrewRadev committed:
Cache a list of authors rather than a citation

AndrewRadev committed:
Author search test

AndrewRadev committed:
Remove unused code

AndrewRadev committed:
Unicode matching fix

AndrewRadev committed:
Don't show the author cache in the admin form

AndrewRadev committed:
Linter fixes

AndrewRadev committed:
Fix admin column name rendering

AndrewRadev committed:
Render page visit stats as a table

AndrewRadev committed:
Merge branch 'page-visit-tracking' into study-authors

* page-visit-tracking:
  Render page visit stats as a table
  Fix admin column name rendering

AndrewRadev committed:
Skip bot visits in table

AndrewRadev committed:
Merge branch 'page-visit-tracking' into study-authors

* page-visit-tracking:
  Skip bot visits in table

AndrewRadev committed:
Search page smoke test

AndrewRadev committed:
Better loading animation on search page

AndrewRadev committed:
Improve search for help pages as well

AndrewRadev committed:
Extract HelpTopics to a separate file

AndrewRadev committed:
Add additional links on help pages

AndrewRadev committed:
How to cite

AndrewRadev committed:
Tweaks to "additional help" section

AndrewRadev committed:
Code comment

AndrewRadev committed:
Initialize crawler detector once

AndrewRadev committed:
Merge branch 'page-visit-tracking' into study-authors

* page-visit-tracking:
  Initialize crawler detector once
  Code comment

AndrewRadev committed:
Some cleanup of HelpTopics

No need to directly depend on `current_app`, or to hardcode the
template paths.

AndrewRadev committed:
Add master thesis links to help page

AndrewRadev committed:
Record errors in production

AndrewRadev committed:
Remove leftover debug error

AndrewRadev committed:
Merge pull request #87 from AndrewRadev/sandbox-page

Sandbox page

AndrewRadev committed:
Merge pull request #88 from AndrewRadev/page-visit-tracking

Page visit counting

AndrewRadev committed:
Merge pull request #90 from AndrewRadev/study-authors

Add authorship fields to studies

AndrewRadev committed:
Fix typo

AndrewRadev committed:
WIP author-fetching

AndrewRadev committed:
Extract crossref api interaction into a model class

AndrewRadev committed:
Fetch real author information, save it

AndrewRadev committed:
Fix author issues for existing studies

AndrewRadev committed:
Error handling

AndrewRadev committed:
Error handling tests

AndrewRadev committed:
Fix compare button bugs

AndrewRadev committed:
Merge branch 'main' into upload-process-improvements-2

* main:
  Fix compare button bugs

AndrewRadev committed:
Reorganize step 1 input fields, add highlighting

AndrewRadev committed:
Show text preview when writing descriptions

AndrewRadev committed:
Swap study/project update functions, fix minor bug

AndrewRadev committed:
Add a link between a study and its last submission

AndrewRadev committed:
Reorganize submission pages to take a submission id

This should make things much simpler instead of juggling the session.
Probably still need to fix some things, though.

AndrewRadev committed:
Extract study choice outside of submission interface

Simplifies a lot of other things if we just pin what study (existing or
new) the form is focused on and we can't change it halfway through.

AndrewRadev committed:
Measure how long it takes to track page visits

AndrewRadev committed:
Record time measurements as DEBUG logs

AndrewRadev committed:
Record time measurements as WARN logs

AndrewRadev committed:
Track API visits separately from web ones

AndrewRadev committed:
Allow filtering page visits by path in the admin

AndrewRadev committed:
Use non-deprecated logging method

AndrewRadev committed:
Merge branch 'main' into upload-process-improvements-2

* main:
  Use non-deprecated logging method
  Allow filtering page visits by path in the admin
  Track API visits separately from web ones
  Record time measurements as WARN logs
  Record time measurements as DEBUG logs
  Measure how long it takes to track page visits

AndrewRadev committed:
Fix favicon metadata tag

AndrewRadev committed:
Merge branch 'main' into upload-process-improvements-2

* main:
  Fix favicon metadata tag

AndrewRadev committed:
Remove usage of SubmissionForm.type, clarify load/create distinction

AndrewRadev committed:
Only show user submissions that are in-progress

AndrewRadev committed:
Hide additional inputs depending on the submission

AndrewRadev committed:
Fix link

AndrewRadev committed:
Add existing study information on status page

AndrewRadev committed:
Minor improvement to study status

AndrewRadev committed:
Record page visits through a background job

AndrewRadev committed:
Move maxminddb out to separate initializer, remove timing information

AndrewRadev committed:
Pin more specific python version for tests

AndrewRadev committed:
Add redis to the github actions setup

AndrewRadev committed:
Return a flag about whether the bioreplicate is an "averaged" one

AndrewRadev committed:
Documentation for the "isAverage" field

AndrewRadev committed:
Allow building documentation from this branch

AndrewRadev committed:
Merge branch 'main' into upload-process-improvements-2

* main:
  Allow building documentation from this branch
  Documentation for the "isAverage" field
  Return a flag about whether the bioreplicate is an "averaged" one
  Add redis to the github actions setup
  Pin more specific python version for tests
  Move maxminddb out to separate initializer, remove timing information
  Record page visits through a background job

AndrewRadev committed:
Fix creating an "update" submission

AndrewRadev committed:
Link newly-finalized submission and study

AndrewRadev committed:
Add an animated loader to upload steps

AndrewRadev committed:
Experiment previews for entered data

AndrewRadev committed:
Minor CSS fix

AndrewRadev committed:
Reorganize technique forms

AndrewRadev committed:
Fix bug when changing ordering of data and forms

AndrewRadev committed:
Add an explanatory note for a tricky part of the template

AndrewRadev committed:
Wording change

AndrewRadev committed:
Don't preview forms with errors, fix ajax form bugs

AndrewRadev committed:
Fix potential issues when building new jQuery elements

AndrewRadev committed:
Save changes when updating an experiment

AndrewRadev committed:
Extract common code, fix issue with duplicate button

AndrewRadev committed:
Require users to submit a description of their changes

AndrewRadev committed:
Fix bug with spreadsheet preview endpoint

AndrewRadev committed:
Fix non-unique duplicate buttons

AndrewRadev committed:
Fix for issues with new submissions

AndrewRadev committed:
Minor reordering and a test

AndrewRadev committed:
Convert measurement context endpoint

AndrewRadev committed:
Convert units in bioreplicate CSVs

AndrewRadev committed:
Implement correct unit conversion for metabolites

If they don't have a mass, in the UI, we don't convert them. We should
do the same in the API.

AndrewRadev committed:
Update API docs

AndrewRadev committed:
Minor doc fix, build documentation from this branch

AndrewRadev committed:
Fix experiment preview state on duplicate

AndrewRadev committed:
Don't minify in development, remove source map URLs

We don't actually serve the source maps, and they generate warnings in
the console.

Additionally, plotly and katex were not being minified in production,
oops.

AndrewRadev committed:
Implement an annoyingly complicated setup for uniqueness errors

For now, this works specifically for the experiment form, though I
should look into the others.

AndrewRadev committed:
Minor label change

AndrewRadev committed:
Better handling of scrolling and tooltips, remove form icons

AndrewRadev committed:
Update form titles

AndrewRadev committed:
Leave note about Performance API

AndrewRadev committed:
Various fixes

AndrewRadev committed:
Merge pull request #94 from AndrewRadev/api-units

Unit conversion in API endpoints

AndrewRadev committed:
Merge branch 'main' into upload-process-improvements-2

* main:
  Minor doc fix, build documentation from this branch
  Update API docs
  Implement correct unit conversion for metabolites
  Convert units in bioreplicate CSVs
  Convert measurement context endpoint

AndrewRadev committed:
Replace unminified plotly.js with minified one

AndrewRadev committed:
Replace plotly 3.3.1 with plotly-basic 3.4.0

AndrewRadev committed:
Use correct plotly path

AndrewRadev committed:
Reword embargo date help message

AndrewRadev committed:
Extract study export to a background job

AndrewRadev committed:
Fix test

AndrewRadev committed:
Allow using days as time units

AndrewRadev committed:
Automatically publish studies after their embargo date has passed

AndrewRadev committed:
Publish button on study page

AndrewRadev committed:
Hide private token from screen, adjust manager-targeted messages

AndrewRadev committed:
Fix leftover debug clause

AndrewRadev committed:
Show long technique names in search results

No point in shortening them there, since they're just one line now

AndrewRadev committed:
Fix bug with published studies

AndrewRadev committed:
Remove time limit from upload forms' CSRF tokens

These forms could potentially take some time and deliberation.

AndrewRadev committed:
Fix bug with compare buttons

The buttons were not initialized correctly on page load, since the data
key changed.

AndrewRadev committed:
Document submissions in the "Main Concepts" help topic

AndrewRadev committed:
Handle a rare potential error if the session and database don't align

AndrewRadev committed:
Fix issue with missing tooltips in nested js forms

AndrewRadev committed:
Move tooltips into separate info icons

AndrewRadev committed:
Update search form screenshots in help topics

AndrewRadev committed:
Update instructions and images in help topics

AndrewRadev committed:
Merge pull request #91 from AndrewRadev/upload-process-improvements-2

Upload process improvements (batch 2)

AndrewRadev committed:
Pin werkzeug to avoid potential issues in tests

AndrewRadev committed:
Release notes for v1.1.0

AndrewRadev committed:
Release notes for v1.1.0

AndrewRadev committed:
Fixes to compartment indexing and name updates

AndrewRadev committed:
Fix issue when sharing studies

If we only give permission to the study and not to the project, it's
impossible to update the project in a follow-up submission.

AndrewRadev committed:
Validate that time points are unique

AndrewRadev committed:
Add "history" page for a study's submissions

AndrewRadev committed:
Make changelogText optional for admin

AndrewRadev committed:
Publish submissions along with studies

AndrewRadev committed:
Add units to 16s measurement message

AndrewRadev committed:
Don't show delete buttons for studies that aren't published yet

AndrewRadev committed:
Show a help message in the preview area

AndrewRadev committed:
Update S1 data spreadsheet

hariszaf commented on README.md:
````diff
@@ -9,13 +9,47 @@ In this repo, you can find the code of the resource, known issues and discussion
+bin/migrations-run
+```
+
+## Local installation for development (on Linux or macOS)
````
If I am not mistaken, you would need Docker both for "local installation using docker" and "local installation for development".
If so, consider mentioning it on the title for being consistent.

AndrewRadev replied:
You don't need docker for mysql and redis, the docker image is provided for convenience if you don't already have them running. The application only needs ports for mysql and and redis, but they could be running in any way.

I've reworded the title to make it clearer that one option is for the entire app inside the docker container, in commit https://github.com/msysbio/bacterial_growth/pull/39/changes/0dbabaafc5b8a6ad02a711e3986004d6f9c8805b

hariszaf commented on README.md:
```diff
@@ -9,13 +9,47 @@ In this repo, you can find the code of the resource, known issues and discussion
+
+If you'd like to get the app running to use in a local environment, a docker container can be a working cross-platform choice. The provided Dockerfile is meant for launching the app in production mode and not for development purposes.
+
+First, copy the `.env.example` file to an `.env` file and fill in the environment variables. Ideally, you should set a random string for session encryption. If you want to be able to log in, you'll need to follow [the ORCID documentation](https://info.orcid.org/documentation/integration-guide/) to create an ORCID app and fill in its information. If you skip that part, you'll only have read access ...
```
This is a bit unclear to me at least. 
Could you mention something more specific on how one can come up with the required secret key etc?

AndrewRadev replied:
I've applied some updates in this commit: https://github.com/msysbio/bacterial_growth/pull/39/changes/0dbabaafc5b8a6ad02a711e3986004d6f9c8805b

hariszaf commented on README.md:
```diff
@@ -9,13 +9,47 @@ In this repo, you can find the code of the resource, known issues and discussion
 - Upload process: <https://mgrowthdb.gbiomed.kuleuven.be/help/upload-process/>
 
-## Local installation
+## Local installation using docker
```
Trying to follow instructions here I got

```
bacterial_growth-mgrowthdb_app-1  | [2026-03-31 11:30:57 +0000] [13] [DEBUG] GET /
bacterial_growth-mgrowthdb_app-1  | [2026-03-31 11:30:57,446] ERROR in app: Exception on / [GET]
bacterial_growth-mgrowthdb_app-1  | Traceback (most recent call last):
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 143, in __init__
bacterial_growth-mgrowthdb_app-1  |     self._dbapi_connection = engine.raw_connection()
bacterial_growth-mgrowthdb_app-1  |                              ^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 3317, in raw_connection
bacterial_growth-mgrowthdb_app-1  |     return self.pool.connect()
bacterial_growth-mgrowthdb_app-1  |            ^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 448, in connect
bacterial_growth-mgrowthdb_app-1  |     return _ConnectionFairy._checkout(self)
bacterial_growth-mgrowthdb_app-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 1272, in _checkout
bacterial_growth-mgrowthdb_app-1  |     fairy = _ConnectionRecord.checkout(pool)
bacterial_growth-mgrowthdb_app-1  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 712, in checkout
bacterial_growth-mgrowthdb_app-1  |     rec = pool._do_get()
bacterial_growth-mgrowthdb_app-1  |           ^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/impl.py", line 177, in _do_get
bacterial_growth-mgrowthdb_app-1  |     with util.safe_reraise():
bacterial_growth-mgrowthdb_app-1  |          ^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py", line 121, in __exit__
bacterial_growth-mgrowthdb_app-1  |     raise exc_value.with_traceback(exc_tb)
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/impl.py", line 175, in _do_get
bacterial_growth-mgrowthdb_app-1  |     return self._create_connection()
bacterial_growth-mgrowthdb_app-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 389, in _create_connection
bacterial_growth-mgrowthdb_app-1  |     return _ConnectionRecord(self)
bacterial_growth-mgrowthdb_app-1  |            ^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 674, in __init__
bacterial_growth-mgrowthdb_app-1  |     self.__connect()
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 900, in __connect
bacterial_growth-mgrowthdb_app-1  |     with util.safe_reraise():
bacterial_growth-mgrowthdb_app-1  |          ^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py", line 121, in __exit__
bacterial_growth-mgrowthdb_app-1  |     raise exc_value.with_traceback(exc_tb)
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 896, in __connect
bacterial_growth-mgrowthdb_app-1  |     self.dbapi_connection = connection = pool._invoke_creator(self)
bacterial_growth-mgrowthdb_app-1  |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/create.py", line 667, in connect
bacterial_growth-mgrowthdb_app-1  |     return dialect.connect(*cargs_tup, **cparams)
bacterial_growth-mgrowthdb_app-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 630, in connect
bacterial_growth-mgrowthdb_app-1  |     return self.loaded_dbapi.connect(*cargs, **cparams)  # type: ignore[no-any-return]  # NOQA: E501
bacterial_growth-mgrowthdb_app-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/pymysql/connections.py", line 365, in __init__
bacterial_growth-mgrowthdb_app-1  |     self.connect()
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/pymysql/connections.py", line 681, in connect
bacterial_growth-mgrowthdb_app-1  |     self._request_authentication()
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/pymysql/connections.py", line 980, in _request_authentication
bacterial_growth-mgrowthdb_app-1  |     auth_packet = _auth.caching_sha2_password_auth(self, auth_packet)
bacterial_growth-mgrowthdb_app-1  |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/pymysql/_auth.py", line 245, in caching_sha2_password_auth
bacterial_growth-mgrowthdb_app-1  |     pkt = conn._read_packet()
bacterial_growth-mgrowthdb_app-1  |           ^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/pymysql/connections.py", line 782, in _read_packet
bacterial_growth-mgrowthdb_app-1  |     packet.raise_for_error()
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/pymysql/protocol.py", line 219, in raise_for_error
bacterial_growth-mgrowthdb_app-1  |     err.raise_mysql_exception(self._data)
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/pymysql/err.py", line 150, in raise_mysql_exception
bacterial_growth-mgrowthdb_app-1  |     raise errorclass(errno, errval)
bacterial_growth-mgrowthdb_app-1  | pymysql.err.OperationalError: (1044, "Access denied for user 'bacterial_growth'@'%' to database 'BacterialGrowth'")
bacterial_growth-mgrowthdb_app-1  | 
bacterial_growth-mgrowthdb_app-1  | The above exception was the direct cause of the following exception:
bacterial_growth-mgrowthdb_app-1  | 
bacterial_growth-mgrowthdb_app-1  | Traceback (most recent call last):
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/flask/app.py", line 1511, in wsgi_app
bacterial_growth-mgrowthdb_app-1  |     response = self.full_dispatch_request()
bacterial_growth-mgrowthdb_app-1  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/flask/app.py", line 919, in full_dispatch_request
bacterial_growth-mgrowthdb_app-1  |     rv = self.handle_user_exception(e)
bacterial_growth-mgrowthdb_app-1  |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/flask/app.py", line 915, in full_dispatch_request
bacterial_growth-mgrowthdb_app-1  |     rv = self.preprocess_request()
bacterial_growth-mgrowthdb_app-1  |          ^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/flask/app.py", line 1291, in preprocess_request
bacterial_growth-mgrowthdb_app-1  |     rv = self.ensure_sync(before_func)()
bacterial_growth-mgrowthdb_app-1  |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/home/initialization/global_handlers.py", line 75, in _open_db_connection
bacterial_growth-mgrowthdb_app-1  |     g.db_conn = get_connection()
bacterial_growth-mgrowthdb_app-1  |                 ^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/home/db/__init__.py", line 70, in get_connection
bacterial_growth-mgrowthdb_app-1  |     return DB.connect()
bacterial_growth-mgrowthdb_app-1  |            ^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 3293, in connect
bacterial_growth-mgrowthdb_app-1  |     return self._connection_cls(self)
bacterial_growth-mgrowthdb_app-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 145, in __init__
bacterial_growth-mgrowthdb_app-1  |     Connection._handle_dbapi_exception_noconnection(
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 2448, in _handle_dbapi_exception_noconnection
bacterial_growth-mgrowthdb_app-1  |     raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 143, in __init__
bacterial_growth-mgrowthdb_app-1  |     self._dbapi_connection = engine.raw_connection()
bacterial_growth-mgrowthdb_app-1  |                              ^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 3317, in raw_connection
bacterial_growth-mgrowthdb_app-1  |     return self.pool.connect()
bacterial_growth-mgrowthdb_app-1  |            ^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 448, in connect
bacterial_growth-mgrowthdb_app-1  |     return _ConnectionFairy._checkout(self)
bacterial_growth-mgrowthdb_app-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 1272, in _checkout
bacterial_growth-mgrowthdb_app-1  |     fairy = _ConnectionRecord.checkout(pool)
bacterial_growth-mgrowthdb_app-1  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 712, in checkout
bacterial_growth-mgrowthdb_app-1  |     rec = pool._do_get()
bacterial_growth-mgrowthdb_app-1  |           ^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/impl.py", line 177, in _do_get
bacterial_growth-mgrowthdb_app-1  |     with util.safe_reraise():
bacterial_growth-mgrowthdb_app-1  |          ^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py", line 121, in __exit__
bacterial_growth-mgrowthdb_app-1  |     raise exc_value.with_traceback(exc_tb)
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/impl.py", line 175, in _do_get
bacterial_growth-mgrowthdb_app-1  |     return self._create_connection()
bacterial_growth-mgrowthdb_app-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 389, in _create_connection
bacterial_growth-mgrowthdb_app-1  |     return _ConnectionRecord(self)
bacterial_growth-mgrowthdb_app-1  |            ^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 674, in __init__
bacterial_growth-mgrowthdb_app-1  |     self.__connect()
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 900, in __connect
bacterial_growth-mgrowthdb_app-1  |     with util.safe_reraise():
bacterial_growth-mgrowthdb_app-1  |          ^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py", line 121, in __exit__
bacterial_growth-mgrowthdb_app-1  |     raise exc_value.with_traceback(exc_tb)
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 896, in __connect
bacterial_growth-mgrowthdb_app-1  |     self.dbapi_connection = connection = pool._invoke_creator(self)
bacterial_growth-mgrowthdb_app-1  |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/create.py", line 667, in connect
bacterial_growth-mgrowthdb_app-1  |     return dialect.connect(*cargs_tup, **cparams)
bacterial_growth-mgrowthdb_app-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 630, in connect
bacterial_growth-mgrowthdb_app-1  |     return self.loaded_dbapi.connect(*cargs, **cparams)  # type: ignore[no-any-return]  # NOQA: E501
bacterial_growth-mgrowthdb_app-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/pymysql/connections.py", line 365, in __init__
bacterial_growth-mgrowthdb_app-1  |     self.connect()
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/pymysql/connections.py", line 681, in connect
bacterial_growth-mgrowthdb_app-1  |     self._request_authentication()
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/pymysql/connections.py", line 980, in _request_authentication
bacterial_growth-mgrowthdb_app-1  |     auth_packet = _auth.caching_sha2_password_auth(self, auth_packet)
bacterial_growth-mgrowthdb_app-1  |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/pymysql/_auth.py", line 245, in caching_sha2_password_auth
bacterial_growth-mgrowthdb_app-1  |     pkt = conn._read_packet()
bacterial_growth-mgrowthdb_app-1  |           ^^^^^^^^^^^^^^^^^^^
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/pymysql/connections.py", line 782, in _read_packet
bacterial_growth-mgrowthdb_app-1  |     packet.raise_for_error()
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/pymysql/protocol.py", line 219, in raise_for_error
bacterial_growth-mgrowthdb_app-1  |     err.raise_mysql_exception(self._data)
bacterial_growth-mgrowthdb_app-1  |   File "/usr/local/lib/python3.12/site-packages/pymysql/err.py", line 150, in raise_mysql_exception
bacterial_growth-mgrowthdb_app-1  |     raise errorclass(errno, errval)
bacterial_growth-mgrowthdb_app-1  | sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (1044, "Access denied for user 'bacterial_growth'@'%' to database 'BacterialGrowth'")
bacterial_growth-mgrowthdb_app-1  | (Background on this error at: https://sqlalche.me/e/20/e3q8)
bacterial_growth-mgrowthdb_app-1  | [2026-03-31 11:30:57,453] WARNING in global_handlers: Couldn't record error in the database (500 Internal Server Error: The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.) due to: current_user
```

Is it my bad or is anything missing from the documentation?

AndrewRadev replied:
Hm, strange, for the docker example, in theory the mysql install script should handle this. I'll try to replicate it and see what I can figure out.

AndrewRadev replied:
I see the problem now. The issue is that the instructions need to be followed to the end, the database needs to be created first, and only then the app can be used. However, because there's now some code that triggers background jobs on page visits (`record_page_visit`), that one tries to access the database and starts throwing errors in the log *before* you've had the chance to run `./init.sh`. (Though I'm not sure why the page visits get triggered, actually. Also something to debug.)

Once again, we have an ordering problem, where the app and its data need to be bootstrapped before the app is launched, but docker does not allow separating the "building" part and the "running" part. I'll give it some thought how to avoid the issue.

hariszaf replied:
OK, do you think we should on with merging the release and see later how this goes?

AndrewRadev replied:
No, I'll figure out a workaround, I have some idea of what to try. I'll write back when I've pushed an update.

hariszaf replied:
:+1:

AndrewRadev replied:
Update: I couldn't fix it properly 😅. I ended up adding a line in the documentation that says not to worry about it yet and first run the "init" script: https://github.com/AndrewRadev/mgrowthdb/commit/c6a84c33918c4e4c8f71e488f828d7923f490e50

I also had to fix an issue where the dockerized application wasn't serving static files. Which I was surprised by, because I feel like this used to work. I couldn't figure out what setting I changed, so I ended up adding nginx to the docker compose setup in the "full" scenario (https://github.com/AndrewRadev/mgrowthdb/commit/17b7814bc23da990c4c884f737aaf40b98ced3a0).

hariszaf reviewed (commented):
I have some very first comments so I can actually give it a go first.

AndrewRadev committed:
Work-in-progress data dashboard with API access

AndrewRadev committed:
Merge pull request #95 from AndrewRadev/upload-process-improvements-3

Upload process improvements (batch 3)

AndrewRadev committed:
Installation clarifications

AndrewRadev committed:
Apply fixes for "full" docker setup, update README

AndrewRadev committed:
Add nginx config

AndrewRadev committed:
Add the youtube channel to the help documentation

AndrewRadev committed:
Allow both chart checkboxes to be unticked

AndrewRadev committed:
Visually fade "blank" trace rows

AndrewRadev committed:
Group measurement contexts by units

AndrewRadev committed:
Use converted units for chart labels

AndrewRadev committed:
Add model prediction group, fix bug

AndrewRadev committed:
Allow downloading a CSV with model predictions

AndrewRadev committed:
Experimental error bands for model predictions

It's a very hacky setup and it's quite frustrating that plotly don't
have a built-in way to implement this. Some problems:

- Colors look more or less reasonable, but it's unclear how they're
  selected. Probably best to implement color palette selection
  ourselves, finally.
- Double-clicking on a trace to isolate it does not remove the error
  band, it just sits there.

AndrewRadev committed:
Remove widths for legend items to let them wrap

I can't remember why I originally introduced this, though it might have
been for a vertical legend. Have to monitor to see if anything breaks.

AndrewRadev committed:
Select/deselect all buttons, scroll shadows

AndrewRadev committed:
Break up long function signatures in docs

AndrewRadev committed:
Remove information paragraph on visualize page, add a "Help" button

AndrewRadev committed:
Rename js-technique-row to the more generic js-trace-row

AndrewRadev committed:
Experimental "subject" filter on visualize page

AndrewRadev committed:
Inline measurement context macro, too restrictive

The contents of the label need to be modifiable, this could
theoretically be achieved with a macro, but it would result in too much
noise.

AndrewRadev committed:
Draw colors explicitly so we can set colors on bands

AndrewRadev committed:
Change grouping label

AndrewRadev committed:
Remove unnecessary fields from MeasurementTechnique

AndrewRadev committed:
Filter secondary dropdowns based on experiment, apply related fixes

AndrewRadev committed:
Minor relabeling

AndrewRadev committed:
Tab styling, restore information paragraph

AndrewRadev committed:
Experimental relabeling of μGrowthDB -> mGrowthDB

AndrewRadev committed:
Documentation

AndrewRadev committed:
Smaller chart controls on comparison page

AndrewRadev committed:
Grouping in compare view

AndrewRadev committed:
Merge pull request #96 from AndrewRadev/chart-improvements

Chart improvements

AndrewRadev committed:
Avoid potential problem with f-string

AndrewRadev committed:
Fixes to RTD config file

AndrewRadev committed:
Fix RTD configuration setting

AndrewRadev committed:
Update documentation links to point to "latest"

AndrewRadev committed:
Modeling API endpoints and documentation

AndrewRadev committed:
Merge pull request #97 from AndrewRadev/remove-unicode-mu-from-text

Experimental relabeling of μGrowthDB -> mGrowthDB

AndrewRadev committed:
Fix bootstrap data for S2

AndrewRadev committed:
Allow export for to export data with STD values

AndrewRadev committed:
ADd a test for `includeStd` flag

AndrewRadev committed:
Use icons instead of emoji, group buttons

AndrewRadev committed:
Update S2 bootstrap measurement data

AndrewRadev committed:
Merge branch 'api-dashboards' into sandbox-page-2

* api-dashboards:
  Work-in-progress data dashboard with API access

AndrewRadev committed:
Relabel sections

AndrewRadev committed:
Reorganize sandbox pages, rename to "workspaces"

AndrewRadev committed:
Reorganize workspace form

AndrewRadev committed:
Fix (temporary) dashboard tests

AndrewRadev committed:
Shorter sidebar label

AndrewRadev committed:
Rename "dashboard entries" to "workspace entries"

AndrewRadev committed:
Add more fields to workspace entries, show them in form

AndrewRadev committed:
Data preview

AndrewRadev committed:
Construct workspace entry records

AndrewRadev committed:
Working data upload

AndrewRadev committed:
Render workspace entries with placeholder buttons

AndrewRadev committed:
Add published state to workspace entries

AndrewRadev committed:
Fix circular dependency error when publishing

AndrewRadev committed:
Merge branch 'main' into sandbox-page-2

* main:
  Fix circular dependency error when publishing
  Update S2 bootstrap measurement data

AndrewRadev committed:
Basic chart form skeleton

AndrewRadev committed:
Group workspace entries under a "Workspace" model

AndrewRadev committed:
Workspace "visualize" page

AndrewRadev committed:
Show a "value"/"error" label on data preview columns

AndrewRadev committed:
Fix tests

AndrewRadev committed:
Error handling

AndrewRadev committed:
Workspace entry info, delete actions

AndrewRadev committed:
Minor CSS fix

AndrewRadev committed:
Move visualization partials

AndrewRadev committed:
Edit entry metadata

AndrewRadev committed:
Fix visualize link, remove comparison button

AndrewRadev committed:
Show "compare" entries in workspaces, unify chart axis checkboxes

AndrewRadev committed:
Minor reordering

AndrewRadev committed:
Group comparable records by study

AndrewRadev committed:
Work-in-progress modeling tab

AndrewRadev committed:
Fix chart placeholder CSS by making it a "utility" class

AndrewRadev committed:
Place description text under tabs to avoid misaligning them

AndrewRadev committed:
Allow modeling results to target workspace entries

AndrewRadev committed:
Partially working modeling chart

AndrewRadev committed:
Modeling form fixes

AndrewRadev committed:
Show modeling results in workspaces

AndrewRadev committed:
Fix test

AndrewRadev committed:
Workspace API endpoints and tests

AndrewRadev committed:
Create, destroy, change workspaces

AndrewRadev committed:
Toggle public/private state

AndrewRadev committed:
Make workspace pages browseable by other users and non-users

AndrewRadev committed:
Select/deselect all buttons, fixes

AndrewRadev committed:
Chart CSS fixes

AndrewRadev committed:
Introduce utility function

AndrewRadev committed:
Minor fixes

AndrewRadev committed:
Unify chart CSS

AndrewRadev committed:
Use "trace-group" instead of "experiment"

AndrewRadev committed:
Add missing margin

AndrewRadev committed:
Fixes

AndrewRadev committed:
Fix migration

AndrewRadev committed:
"Link to this page" improvements

AndrewRadev committed:
Fix visualize page URL

AndrewRadev committed:
Fix visualization page error

AndrewRadev committed:
Minor ORM field fixes

AndrewRadev committed:
Fix issue with submitting model forms without an active radio button

AndrewRadev committed:
Merge pull request #98 from AndrewRadev/sandbox-page-2

Workspace pages (Sandbox, part 2)

AndrewRadev committed:
Add a license URL and license image, whenever available

AndrewRadev committed:
Adjust the homepage information

AndrewRadev committed:
Add full data export link to sidebar menu

AndrewRadev committed:
Allow TOC navigation to overflow and scroll

AndrewRadev committed:
README fix

AndrewRadev committed:
Add FAIRsharing.org link to the docs

AndrewRadev committed:
Add strain & metabolite counts to result tables

AndrewRadev committed:
Fetch publication date and add it to studies

AndrewRadev committed:
Add sort ordering to search page, allow sorting by publication date

AndrewRadev committed:
Show experiment and bioreplicate count in search results

AndrewRadev committed:
Add a README.txt file for the full export

AndrewRadev committed:
Workspaces documentation: direct upload

AndrewRadev committed:
Reformat "README.txt" file for data exports

AndrewRadev committed:
Add youtube tutorial to upload help page

AndrewRadev committed:
More workspaces documentation

AndrewRadev committed:
Workspace pushing through API

AndrewRadev committed:
Empty __init__ file for documentation generation

AndrewRadev committed:
Workspace API documentation

AndrewRadev committed:
Organize user profile page with tabs, reset API key

AndrewRadev committed:
Fix profile page tests

AndrewRadev committed:
More workspace documentation and minor improvements

AndrewRadev committed:
Documentation

AndrewRadev committed:
Make the sidebar logo a link to the homepage

AndrewRadev committed:
Add "Averages" tab to visualize section

AndrewRadev committed:
Add license url to study upon submission

AndrewRadev committed:
Fixes to labeling in "Averages" tab

AndrewRadev committed:
Fix submission validation if mandatory columns are missing

AndrewRadev committed:
Increase "averages" form max height

AndrewRadev committed:
Download export archive via POST

AndrewRadev committed:
Create default workspaces, set API key on registration

AndrewRadev committed:
Fix redirect issue

AndrewRadev committed:
Filter non-numeric columns, show warnings

AndrewRadev committed:
Fix orcid login issues

AndrewRadev committed:
Validate workspace upload time values

AndrewRadev committed:
Don't upload data if there were any errors

AndrewRadev committed:
Create a workspace on API push if it doesn't exist

AndrewRadev committed:
Homepage examples and re-wording

AndrewRadev committed:
Minor rewording

AndrewRadev committed:
Update link to workspaces playlist rather than to first video

AndrewRadev committed:
Fix project name and description upon submission creation

AndrewRadev committed:
Test change for deployment testing

AndrewRadev committed:
Revert "Test change for deployment testing"

This reverts commit b465a16333e3c4a0f37a27357d9aedc9a56512fc.

AndrewRadev committed:
Allow admins to make submissions to studies they don't own

AndrewRadev committed:
Set a maximum embargo date

AndrewRadev committed:
Fix bug with automatic averages creation

AndrewRadev committed:
Allow downloading measurement data for unpublished studies

For the API, providing an apiKey will allow people to download their own
unpublished data. This is not documented yet.

For the web, this will allow downloading data using g.current_user.

AndrewRadev committed:
Adapt averages creation to take the common points

AndrewRadev committed:
Fix author cache problem

AndrewRadev committed:
Delete unused ModelingRequests table

AndrewRadev committed:
Minor CSS tweaks on project pages

AndrewRadev committed:
Add publication type to studies

AndrewRadev committed:
Fix homepage description

AndrewRadev committed:
Set publication type during the upload process

AndrewRadev committed:
Fix bug with setting study title

AndrewRadev committed:
Trigger a 404 on missing experiments

AndrewRadev committed:
Increase gunicorn timeout

hariszaf commented on app/model/lib/crossref_fetcher.py:
```diff
@@ -0,0 +1,64 @@
+    list and the title of a study.
+    """
+    def __init__(self, doi):
+        self.doi = doi
```
Consider having something like:
`doi.strip('https://doi.org/')` just in case ? except if you re making sure of that somewhere else

AndrewRadev replied:
The crossref API works fine if the DOI given is a URL starting with `https://doi.org`, so this works in practice. Though weirdly enough, I can't quite tell where `requests` is escaping it, so maybe I'll add an explicit quoting function just in case.

hariszaf reviewed (commented):
This is a massive PR most of which has been already reviewer, within @AndrewRadev fork. 

I guess it's more than good to go, since it has been already tested in production mode.
